File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        vector<int> counts(26);

        counts[s[0]-'A']++;
        int l = 0, r = 1, rslt = k, curMax = 1;
        while(r < s.size()){
            counts[s[r]-'A']++;

            curMax = max(curMax, counts[s[r]-'A']);

            if(r-l+1-curMax <= k){
                rslt = max(rslt, r-l+1);
            } else{
                while(r-l+1-curMax > k){
                    counts[s[l]-'A']--;
                    l++;
                }
            }

            r++;
        }

        return rslt;
    }
};

File: Data Structures & Algorithms/partition-labels/submission-1.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        lastIndex = dict()

        for i, c in enumerate(s):
            lastIndex[c] = i
        
        rslt = list()
        size = 0
        cur_last = 0

        for i, c in enumerate(s):
            size += 1

            cur_last = max(cur_last, lastIndex[c])

            if i == cur_last:
                rslt.append(size)
                size = 0
        
        return rslt
            